[ONNX] Update back the mutated buffers into the original PyTorch model

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113686
* #112272
* #113495

Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

Commit:
[ONNX] Update back the mutated buffers into the original PyTorch model

Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

thiagocrepaldi commented:
@BowenBao PTAL

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -660,7 +663,48 @@ def __call__(self, *args: Any, options=None, **kwargs: Any) -> Any:
+    def update_mutated_buffers_onto_torch_model(
+        self,
+        torch_model: torch.nn.Module,
+        onnxruntime_outputs: Sequence[
```
```suggestion
        onnx_outputs: Sequence[
```
nit: doesn't have to be runtime specific?

BowenBao reviewed (commented):
What is the motivation? Feels like `ExportOutput` aka `ONNXProgram` now is expanding in size in a considerably fast rate.

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

Commit:
Update on "[ONNX] Update back the mutated buffers into the original PyTorch model"


Currently, when the user_model: torch.export.ExportedProgram is exported
to ONNX and its implementation mutates buffers, the mutated buffer is
returned as an extra output in the ONNX graph while the original model
still have the outdated buffer value.

If we do a numerical comparison between the pytorch model and the ONNX
model, this mismatch is detected, failing the numerical equivalence
check

This PR updates the pytorch model with the new buffer value after the
ONNX model is executed, making both models numerically equivalent.

[ghstack-poisoned]

thiagocrepaldi commented:
Closing in favor of an updated https://github.com/pytorch/pytorch/pull/112272
